### PR TITLE
Ignition messages not correctly decoded

### PR DESCRIPTION
**Pull request by jcardus**
Ignition messages were not correctly decoded when "Position Append Mask" is configured with value 1 to include "GNSS Satellite Number"

I added two test cases that were failing.

https://github.com/traccar/traccar/files/14797521/GV310LAU.%40Track.Air.Interface.Protocol.V2.02.pdf

image

### CODE REVIEW

**Comment by tananaev** [comment]
Please attach the documentation.

File: src/test/java/org/traccar/protocol/Gl200TextProtocolDecoderTest.java [hunk 1]
```diff
@@ -513,6 +513,12 @@ public void testDecode() throws Exception {
         verifyAttributes(decoder, buffer(
                 "+RESP:GTCAN,8020050605,867488060270575,,00,1,FFFFFFFF,8LBETF3W4N0001613,,,22.54,0,,,,,,,7.84,4.61,3.24,3.33,,8080,,,00,0.00,0.00,1,14,14,2371,0,001FFFFF,,,,,,,,,7158,9998,0,7.84,0.00,0.00,558,,,,,,,C0,,,,,0,0.0,346,2848.5,-78.592371,-0.968132,20240202083437,0740,0002,526C,00AE7907,00,20240202083440,3F6D$"));
 
+        verifyAttribute(decoder, buffer(
+                        "+BUFF:GTIGN,6E0202,868589060169789,ra79,379,1,0.0,105,532.2,-70.616413,-33.393457,20240610201712,0730,0001,333A,00CFA301,01,11,,0.0,20240610201713,3AE2$"),
+                Position.KEY_IGNITION, true);
+        verifyAttribute(decoder, buffer(
+                        "+RESP:GTIGF,6E0202,868589060169789,ra79,145,1,0.0,83,532.2,-70.616413,-33.393457,20240610201937,0730,0001,333A,00CFA301,01,12,,0.0,20240610201938,3AE9$"),
+                Position.KEY_IGNITION, false);
```
Comment (tananaev) [inline src/test/java/org/traccar/protocol/Gl200TextProtocolDecoderTest.java:521]: Fix formatting

Comment (jcardus) [reply src/test/java/org/traccar/protocol/Gl200TextProtocolDecoderTest.java]: done

File: src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java [hunk 2]
```diff
@@ -1033,36 +1033,33 @@ private Object decodeEri(Channel channel, SocketAddress remoteAddress, String[]
...
-            .text("$").optional()
-            .compile();
-
-    private Object decodeIgn(Channel channel, SocketAddress remoteAddress, String sentence) {
-        Parser parser = new Parser(PATTERN_IGN, sentence);
-        Position position = initPosition(parser, channel, remoteAddress);
-        if (position == null) {
+    private Object decodeIgn(
+            Channel channel, SocketAddress remoteAddress, String[] v, String type) throws ParseException {
+        int index = 1; // header
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java:1038]:
```suggestion
        int index = 0;
        index += 1; // header
```

File: src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java [hunk 3]
```diff
@@ -1033,36 +1033,33 @@ private Object decodeEri(Channel channel, SocketAddress remoteAddress, String[]
...
-            .text(",")
-            .number("(xxxx)")                    // count number
-            .text("$").optional()
-            .compile();
-
-    private Object decodeIgn(Channel channel, SocketAddress remoteAddress, String sentence) {
-        Parser parser = new Parser(PATTERN_IGN, sentence);
-        Position position = initPosition(parser, channel, remoteAddress);
-        if (position == null) {
+    private Object decodeIgn(
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java:1036]: Please also don't forget to attach the documentation for future reference.

Comment (jcardus) [reply src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java:1036]: documentation attached

**Comment by tananaev** [comment]
Merged, thank you.

File: src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java [hunk 4]
```diff
@@ -1033,36 +1033,34 @@ private Object decodeEri(Channel channel, SocketAddress remoteAddress, String[]
...
 
-        decodeLocation(position, parser);
+        String model = getDeviceModel(deviceSession, protocolVersion);
+        index += 1; // device name
+        index += 1; // duration of ignition on/off
 
-        position.set(Position.KEY_IGNITION, sentence.contains("GN"));
-        position.set(Position.KEY_HOURS, parseHours(parser.next()));
-        position.set(Position.KEY_ODOMETER, parser.nextDouble() * 1000);
+        decodeLocation(position, model, v, index);
```
Comment (tananaev) [inline src/main/java/org/traccar/protocol/Gl200TextProtocolDecoder.java:1052]: This actually broke decoding:

https://www.traccar.org/forums/topic/gl200-protocol-error-in-gv55w-device/

Fixed it here: https://github.com/traccar/traccar/commit/b1e0484ffd8fe8e71b4027cdb06d23431f5dfac8. Please be more careful with changes in the future.

**Pull request merged**